src/celeritas/optical/ScintillationGenerator.hh

```diff
@@ -0,0 +1,206 @@
+    }
+
+  private:
+    //// TYPES ////
+
+    using UniformRealDist = UniformRealDistribution<real_type>;
+    using ExponentialDist = ExponentialDistribution<real_type>;
+
+    //// DATA ////
+
+    OpticalDistributionData const& dist_;
+    ScintillationSpectrum const& spectrum_;
+    Span<OpticalPrimary> photons_;
+
+    UniformRealDist sample_cost_;
+    UniformRealDist sample_phi_;
+    ExponentialDist sample_time_;
+
+    bool is_neutral_{};
+    units::LightSpeed delta_speed_{};
+    Real3 delta_pos_{};
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from shared scintillation data and distribution parameters.
+ */
+CELER_FUNCTION
+ScintillationGenerator::ScintillationGenerator(
+    OpticalDistributionData const& dist,
+    NativeCRef<ScintillationData> const& shared,
+    Span<OpticalPrimary> photons)
+    : dist_(dist)
+    , spectrum_(shared.spectrum[dist_.material])
+    , photons_(photons)
+    , sample_cost_(-1, 1)
+    , sample_phi_(0, 2 * constants::pi)
+    , sample_time_(real_type{1} / spectrum_.fall_time[0])
+    , is_neutral_{dist_.charge == zero_quantity()}
+{
+    CELER_EXPECT(dist_);
+    CELER_EXPECT(photons_.size() == dist_.num_photons);
+
+    auto const& pre_step = dist_.points[StepPoint::pre];
+    auto const& post_step = dist_.points[StepPoint::post];
+    delta_pos_ = post_step.pos - pre_step.pos;
+    delta_speed_ = post_step.speed - pre_step.speed;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample scintillation photons from optical property data and step data.
+ *
+ * The optical photons are generated evenly along the step and are emitted
+ * uniformly over the entire solid angle with a random linear polarization.
+ * The photon energy is calculated by the scintillation emission wavelength
+ * \f[
+   E = \frac{hc}{\lambda},
+ * \f]
+ * where \f$ h \f$ is the Planck constant and \f$ c \f$ is the speed of light,
+ * and \f$ \lambda \f$ is sampled by the normal distribution with the mean of
+ * scintillation emission spectrum and the standard deviation. The emitted time
+ * is simulated according to empirical shapes of the material-dependent
+ * scintillation time structure with one or double exponentials.
+ */
+template<class Generator>
+CELER_FUNCTION Span<OpticalPrimary>
+ScintillationGenerator::operator()(Generator& rng)
+{
+    // Loop for generating scintillation photons
+    size_type num_generated{0};
+
+    for (auto sid : range(spectrum_.size))
+    {
+        // Skip if there is no emission warelength
+        if(spectrum_.lambda_mean[sid] <= 0) 
+           continue;
+
+        NormalDistribution<real_type> sample_lambda(
+            spectrum_.lambda_mean[sid], spectrum_.lambda_sigma[sid]);
+
+        size_type num_photons = dist_.num_photons * spectrum_.yield_prob[sid];
```

Comment:
With this construction, isn't it possible for the total number of photons to be rounded down so that the number of photons is less than `dist_.num_photons`?

Reply:
A good point! I also thought about the better way how to deal with a potential mis-match due to double->int conversion.  A sure solution is to pass an extra num_to_generate[3] as a part of ScintillationParams of which sum should match to dist_.num_photons.  Here, the assumption is that the sum of yield_prob*dist_.num_photons is already adjusted to dist_.num_photons when input dist_ is prepared in the process level.  We can revisit this when ScintillationProcess is implemented.

Reply:
True, we should. I guess then we'd need to make sure enough storage is allocated for a few extra photons, or later use a stack allocator to allocate it inside the generator, unless we move this part to the pre-generator.

Reply:
There's got to be a smart and efficient way to guarantee that we divide N photons fairly into M groups based on the yield probabilities... (the hard part might be to do it without allocating any extra space, and to make sure that all photons with the same spectra are grouped together so we can reuse the gaussian sampler...)

Reply:
And actually if we want to preserve the distribution of photons rather than just the mean, shouldn't we use a binomial distribution to sample the number from each group? (I.e. sometimes we might sample all short-wavelength photons, sometimes all long-wavelength.)

Reply:
It is a good argument whether each component of scintillation spectra is independent of (which should be the case, but hard to related what amount of energy (i.e. from ionization) actually contributes to each scintillation component - in the Geant4 approach, the total number of photons is sampled by the Poisson distribution with the sum of mean number of photons for each component, which again is used to estimate the number of photons to generate for each component based on provided yields (so in some sense the stochastic treatment is already done before the generator.  As I shortly mentioned above,  a clean solution is to estimate all numbers of photons to generate  before the generator and pass them via the distribution (i.e, input data), which prevents the the rounding error from the loop (at the cost of additional data entry in the OpticalDistributionData, but some redundant information (i.e., sum of all components  == num_photons) or make the OpticalDistributionData different between Cerenkov (just need the total number of photons) and Scintillation (need numbers for all components separately). An alternative tweak with keeping current data structures as they are  is something  like (looks ugly) 
```
     // Calculate the number of photons for each scintillation component         
    Array<size_type, ScintillationSpectrum::size> num_photons{};
    size_type remaining = dist_.num_photons;
    for (auto sid : range(spectrum_.size))
    {
        num_photons[sid] = static_cast<size_type>(dist_.num_photons
                                                  * spectrum_.yield_prob[sid]);
        remaining -= num_photons[sid];
    }

    // Add remaining due to the rounding off, if any, to the first component (or can add to the last
    // none-empty yield component with extra codes    
    if (remaining != 0)
    {
        num_photons[0] += remaining;
    }
```
and loop over with all num_photons[3], at then ensure
`  CELER_ASSERT(num_generated == dist_.num_photons);`
after the loop. How do you think?

Reply:
I'm sure there are better ways, but one possibility to divide the number of photons fairly (similar to what @whokion suggested, but I think in an unbiased way) is to precompute the number of photons outside the loop, first calculating the integer part and then using the normalized fractional parts as a PMF to sample which spectra the remaining photon(s) will go to.  Something like:
```c++
Array<size_type, 3> num_photons;
Real3 frac;
real_type frac_sum{0};
size_type count{0};
for (int i = 0; i < 3; ++i) 
{
    real_type num_photons_real = dist_.num_photons * spectrum_.yield_prob[i];
    num_photons[i] = num_photons_real;
    frac[i] = num_photons_real - num_photons[i];
    frac_sum += frac[i];
    count += num_photons[i];
}
for (size_type i = count; i < dist_.num_photons; ++i)
{
    real_type cutoff = frac_sum * generate_canonical(rng);
    real_type frac_cumsum{0};
    for (int j = 0; j < 3; ++j)
    {
        frac_cumsum += frac[j];
        if (frac_cumsum > cutoff) 
        {
            ++num_photons[j];
            break;  
        }
    }
}
```
But maybe this is overkill.

Reply:
As the number of photons is large (~40000/MeV energy loss for LAr, for an example, without a field), and the user provided yields is a rough estimation, the bias (where to assign extra 1, if happens, most of cases) may be really a negligible effect without loss of generality. Thank you for the suggestion anyway!